utils/src/merkle_tree/optimal_merkle_tree.rs

```diff
@@ -316,6 +322,66 @@ where
         }
         Ok(())
     }
+
+    /// Update hashes after some leaves have been set or updated
+    /// index - first leaf index (which has been set or updated)
+    /// length - number of elements set or updated
+    fn update_hashes(&mut self, index: usize, length: usize) -> Result<()> {
+        // parent depth & index (used to store in the tree)
+        let mut parent_depth = self.depth - 1; // tree depth (or leaves depth) - 1
+        let mut parent_index = index >> 1;
+        let mut parent_index_bak = parent_index;
+        // maximum index at this depth
+        let parent_max_index_0 = (1 << parent_depth) / 2;
+        // Based on given length (number of elements we will update)
+        // we could restrict the parent_max_index
+        let current_index_max = if (index + length) % 2 == 0 {
+            index + length + 2
+        } else {
+            index + length + 1
+        };
+        let mut parent_max_index = min(current_index_max >> 1, parent_max_index_0);
+
+        // current depth & index (used to compute the hash)
+        // current depth initially == tree depth (or leaves depth)
+        let mut current_depth = self.depth;
+        let mut current_index = if index % 2 == 0 { index } else { index - 1 };
+        let mut current_index_bak = current_index;
+
+        loop {
+            // println!("=====");
+            // println!("parent depth: {}, index: {}", parent_depth, parent_index);
+            // println!("parent max index: {}", parent_max_index);
+            // println!("current depth: {}, index: {}", current_depth, current_index);
+
+            // Hash 2 values at (current depth, current_index) & (current_depth, current_index + 1)
+            let n_hash = self.hash_couple(current_depth, current_index);
+            // Insert this hash at (parent_depth, parent_index)
+            self.nodes.insert((parent_depth, parent_index), n_hash);
+
+            if parent_depth == 0 {
+                // We just set the root hash of the tree - nothing to do anymore
+                break;
+            }
+            // Incr parent index
+            parent_index += 1;
+            // Incr current index (+2 because we've just hashed current index & current_index + 1)
+            current_index += 2;
+            if parent_index >= parent_max_index {
+                // reset (aka decr depth & reset indexes)
+                parent_depth -= 1;
+                parent_index = parent_index_bak >> 1;
+                parent_index_bak = parent_index;
+                // parent_max_index = (1 << parent_depth) / 2;
```

Comment:
I don't really understand what this explanation of the code below means

Reply:
Not really a comment but old code. Removed it